flang/lib/Lower/OpenMP.cpp

```diff
@@ -2433,10 +2502,44 @@ genTargetOp(Fortran::lower::AbstractConverter &converter,
                  Fortran::parser::OmpClause::Defaultmap>(
       currentLocation, llvm::omp::Directive::OMPD_target);
 
-  return genOpWithBody<mlir::omp::TargetOp>(
-      converter, eval, currentLocation, outerCombined, &clauseList,
-      ifClauseOperand, deviceOperand, threadLimitOperand, nowaitAttr,
-      mapOperands);
+  auto captureImplicitMap = [&](const Fortran::semantics::Symbol &sym) {
+    if (llvm::find(mapSymbols, &sym) == mapSymbols.end()) {
+      mlir::Value baseOp = converter.getSymbolAddress(sym);
+      if (!baseOp)
+        if (const auto *details = sym.template detailsIf<
+                                  Fortran::semantics::HostAssocDetails>()) {
+          baseOp = converter.getSymbolAddress(details->symbol());
+          converter.copySymbolBinding(details->symbol(), sym);
+        }
+
```

Comment:
You can check here what properties are not constant an pass them too or pass something with more info. With HLFIR, you could also pass the "entity", i.e., the first result of the hlfir.declare: it contains all the required info to work with the symbol (but there is no tool exposed in the bridge to do that, but when we move to HLFIR only, I see no problem adding it).

Reply:
Hi, Thanks a lot for explaining the source of the error.

I am however struggling to code the solution for this as I am unfamiliar with the PFT lowering. Can you please tell me how to get the properties here and check if they are constant?

Also, I was able to add the extents as block arguments and it worked, but how do I get the original symbol(`Fortran::symantics::symbol`) for the extent values(`mlir::values`)? I need the symbols to bind them.

And is there any existing code which does something similar which I can use as reference? I wasn't able to find any.

Reply:
Hi @kiranchandramohan and others, I'd welcome any suggestions that you have for me here.

Reply:
Hi @TIFitis, first of all you cannot map extents/lengths to symbols: specifications in Fortran can be many other things than named objects (something that has a semantics::Symbols). An extent may be, `2*n+1`, there is no symbol for that.

You do not need to care about PFT/Symbols IMHO. You should most likely start from the ExtendedValue SSA values instead of going back to the Symbol evaluate::Expr (this is non trivial and there would be quite some risk to miss some relevant specification value).

My advice:

Either 1.:
- Manage to make a single value for the whole entity being passed as a block argument (i.e, make it a fir.boxchar/fir.box when some extents/lengths are needed). This is easier to do with HLFIR, since symbols are mapped to single SSA value in HLFIR. Probably the cleanest, but may not lead to performant code right away (too much "opaque" fir.box).

Or 2.:
- make some fir::ExtendedValue visitor that visits all the SSA value.
- You call it a first time when creating your block argument types, for non constant SSA values (can checked with something like [this helper that you could expose in FIRBuilder.h](https://github.com/llvm/llvm-project/blob/22f81b4cf4cbf001a8a27877274eb157a12466c3/flang/include/flang/Optimizer/Dialect/FIROpsSupport.h#L136)), push a block arguments.
- You call it a second time when mapping the new symbol inside the block, to re-create the constant value and map the block arguments/new ones (assuming the visit is non const and you can modify the SSA values on the fly).

Or 3.:
- You lower "as usual" in some not isolated from above region, and run some MLIR based pass later to capture SSA values used from outside via IRMapping and cloning after the target region is lowered. You should most likely do it in a pass after lowering (advised for better testing/debugging). It may just be the easiest IMHO.

Reply:
I have a feeling 3 is perhaps similar to what the EarlyOutliingPass currently does: https://github.com/llvm/llvm-project/blob/e18dca257f81ab8c4d00924b3f403a8dddfdddca/flang/lib/Optimizer/Transforms/OMPEarlyOutlining.cpp#L25 but perhaps I'm wrong, not a bad thing but it is a method we already have in place so it'd be interesting to know if it was a similar approach to 3.

Reply:
Yes, 3. is likely very similar in logic to early outlining, expect you "outline" to a nested isolated op instead of a func.func.

Reply:
@TIFitis Which approach are you considering?

Reply:
@kiranchandramohan I am refraining from approach 3 as that puts us back in a messy situation that we were trying to avoid with this patch.

Currently I'm trying to implement approach 1. I was able to extract all the bounds/extents information and add them as block arguments which seems to be working fine for lowering to FIR/HLFIR.

But, we also need to add them as `map_operands` for lowering from MLIR to LLVM later on. The `mapInfoOp` uses a `Pointer_Like_Type` for it's result type and the bounds/extents are `index` type, this is causing incompatibility. I might change the type to AnyType for the mapInfo related ops for now, and later use the verifier to check type compatibility with OpenMP.

Does this approach seem sound to you and others?